siad/API.md

```diff
@@ -0,0 +1,387 @@
+Siad API
+========
+
+All API calls return JSON objects. If there is an error, the error is returned
+in plaintext. The standard response is { "Success": true}. In this document,
+the API responses are defined as go structs. The go structs will get encoded to
+JSON before being sent. Go structs are used here to provide type information.
+
+At version 0.4, the API will be locked into forwards compatibility. This means
+that we will not remove elements from the responses (but we may add additional
+elements), and we will not add any required parameters (though we may add
+optional parameters, and we may disable parameters).
+
+There may be functional API calls which are not documented. These are not
+guaranteed to be supported beyond the current release, and should not be used
+in production.
+
+Daemon
+------
+
+Queries:
+
+* /daemon/stop
+* /daemon/update/apply
+* /daemon/update/check
+
+#### /daemon/stop
+
+Function: Cleanly shuts down the daemon. May take a while.
+
+Parameters: none
+
+Response: standard
+
+#### /daemon/update/apply:
+
+Function: Applies any updates that are available.
+
+Parameters: none
+
+Response: standard
+
+#### /daemon/update/check:
+
+Function: Checks for an update, returning a bool indicating whether
+there is an update and a version indicating the version of the update.
+
+Parameters: none
+
+Response:
+```
+struct {
+	Available bool
+	Version   string
+}
+```
+
+Consensus
+---------
+
+Queries:
+
+* /consensus/status
+
+#### /consensus/status
+
+Function: Returns information about the consensus set, such as the current
+block height.
+
+Parameters: none
+
+Response:
+```
+struct {
+	Height       int
+	CurrentBlock string
+	Target       string
+}
+```
+
+Gateway
+-------
+
+Queries:
+
+* /gateway/status
+* /gateway/synchronize
+* /gateway/peer/add
+* /gateway/peer/remove
+
+#### /gateway/status
+
+Function: Returns information about the gateway, including the list of peers.
+
+Parameters: none
+
+Response:
+```
+struct {
+	Peers []string
+}
+```
+
+#### /gateway/synchronize
+
+Function: Will synchronize the daemon + consensus to the rest of the network.
+Effects may take a few minutes to set in. Should only be necessary for
+debugging.
+
+Parameters: none
+
+Reponse: none
+
+#### /gateway/peer/add
+
+Function: Will add a peer to the gateway.
+
+Parameters:
+```
+struct {
+	Address string
+}
+```
+Address should be an IP address of the form: "a.b.c.d:port"
```

Comment:
technically there are no such restrictions; all that matters is that the host is pingable. Perhaps more accurate would be:
```
Address should be a reachable IP address or hostname, along with a port number. Examples:
    123.123.123.123:9001
    [2001:0DB8:AC10:FE01::]:9001
    domain.example:9001
```

same goes for `remove`